### PR TITLE
Fixed several issues to re-enable code coverage

### PR DESCRIPTION
**Pull request by asvrada**
1. Removed the older version of compiler-rt
2. Disabled some conflicting compiler flag when generating code coverage
3. Disableb a few unit tests when checking code coverage
4. Updated tool chain for code coverage

Currently, with the Spectre mitigation and code coverage check enabled, building and running the test cases (`cmake .. -DCODE_COVERAGE=ON -DUSE_DEBUG_MALLOC=OFF -DENABLE_REFMAN=OFF`) would result in program hanging.

This PR also skips/disables several unit tests that would conflict with or fail with code coverage enabled.

This PR also updates the tool chain used: `llvm-cov-10` and `lcov_cobertura`

Signed-off-by: Zijie Wu <zijiewu@microsoft.com>

### CODE REVIEW

File: cmake/compiler_settings.cmake
Comment (mingweishih) [inline cmake/compiler_settings.cmake:66]:
```suggestion
# More specifically, the speculative-load-hardening instrumentation conflicts with
# gcov ones, causing segfaults in some tests
```

File: cmake/compiler_settings.cmake
Comment (mingweishih) [inline cmake/compiler_settings.cmake:65]:
```suggestion
# Do not apply Spectre mitigation when generating code coverage because of a compiler bug.
```

File: 3rdparty/libgcov/CMakeLists.txt
Comment (mingweishih) [inline 3rdparty/libgcov/CMakeLists.txt:4]: We can consider directly building the libgcov as part of compiler-rt folder, avoiding these copy operations.

File: CMakeLists.txt
Comment (mingweishih) [inline CMakeLists.txt:341]: We should keep the warning message to avoid potential conflicting with other compiler versions. Just update the version.

File: tests/log/host/host.c
Comment (mingweishih) [inline tests/log/host/host.c:27]: Can you double check which point does the test fail? Using SKIP_RETURN_CODE with code coverage is only beneficial if some enclave code can be run (so they can contribute to the code coverage data). If we have to return at this point, we should just disable the test.

**Comment by mingweishih** [comment]
@asvrada Can you work with @CyanDevs to see how can we enable the code-coverage pipeline? It's currently disabled: https://github.com/openenclave/openenclave/blob/e4e7fa21c0a41c2b017872a32ab60f4da84c406c/.jenkins/pipelines/Azure/Linux/Jenkinsfile#L81

**Comment by asvrada** [comment]
> @asvrada Can you work with @CyanDevs to see how can we enable the code-coverage pipeline? It's currently disabled:
> 
> https://github.com/openenclave/openenclave/blob/e4e7fa21c0a41c2b017872a32ab60f4da84c406c/.jenkins/pipelines/Azure/Linux/Jenkinsfile#L81

There is a followup item to enable the pipeline after merging this PR, so we can merge this one and see if everything works then enable the pipeline.

**Comment by mingweishih** [comment]
> > @asvrada Can you work with @CyanDevs to see how can we enable the code-coverage pipeline? It's currently disabled:
> > https://github.com/openenclave/openenclave/blob/e4e7fa21c0a41c2b017872a32ab60f4da84c406c/.jenkins/pipelines/Azure/Linux/Jenkinsfile#L81
> 
> There is a followup item to enable the pipeline after merging this PR, so we can merge this one and see if everything works then enable the pipeline.

We should at least enable the pipeline with code coverage CMake configurations so that all these changes will be tested. We can defer the Jenkins/github integration later (so that you don't need to deal with the dependencies on the front-end tools in the PR).

**Comment by mingweishih** [comment]
bors try

File: tests/log/host/host.c
Comment (asvrada) [reply tests/log/host/host.c]: tests/log has been disabled as a whole so I will remove these code changes. As for the point of failure, this test would output and save text to a file on the file system, and since we override filesystem for code coverage, it fails to find the program's output.

File: tests/log/log.edl
Comment (mingweishih) [inline tests/log/log.edl:6]: nit: drop the unnecessary change

File: tests/log/enc/CMakeLists.txt
Comment (mingweishih) [inline tests/log/enc/CMakeLists.txt:11]: This should be unnecessary now

File: tests/log/CMakeLists.txt
Comment (mingweishih) [inline tests/log/CMakeLists.txt:11]: nit: drop the unnecessary change

**Review by mingweishih — COMMENTED** [review]
Looks good to me overall. Will approve once confirming the pipeline is properly enabled.

**Comment by CyanDevs** [comment]
Enabling the test on Jenkins will take some work to use an alternative method as the plugins used as part of the code coverage test are outdated and has unfixed vulnerabilities. The specific ones are seen in lines 25 and 34 of  https://github.com/openenclave/openenclave/blob/master/.jenkins/library/vars/tests.groovy#L25-L34

**Comment by mingweishih** [comment]
> Enabling the test on Jenkins will take some work to use an alternative method as the plugins used as part of the code coverage test are outdated and has unfixed vulnerabilities. The specific ones are seen in lines 25 and 34 of https://github.com/openenclave/openenclave/blob/master/.jenkins/library/vars/tests.groovy#L25-L34

There also are some extra dependencies such as `lcov` and `lcov_cobertura` (python module) that the image needs to support. We can make the full integration as the follow-up.

@asvrada Let's comment out this line for now.
https://github.com/openenclave/openenclave/blob/65273c3117553139d67f03e21c1fb2adcb2d5d78/.jenkins/library/vars/tests.groovy#L20

File: .jenkins/library/vars/tests.groovy
Comment (mingweishih) [inline .jenkins/library/vars/tests.groovy:17]: You need to also comment out line 24 - 39

**Comment by asvrada** [comment]
bors try

**Review by mingweishih — APPROVED** [review]
LGTM
@CyanDevs please review the Jenkins-related changes.

**Review by CyanDevs — APPROVED** [review]
LGTM

**Comment by mingweishih** [comment]
bors try

**Comment by asvrada** [comment]
bors try

**Pull request merged**
